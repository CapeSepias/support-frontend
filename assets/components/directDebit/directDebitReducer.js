// @flow

// ----- Imports ----- //

import type { Action } from './directDebitActions';

// ----- Setup ----- //

export type DirectDebitState = {
  isPopUpOpen: boolean,
  sortCodeArray: Array<string>,
  bankAccountNumber: string,
  accountHolderName: string,
  accountHolderConfirmation: boolean,
  formError: string
};


const initialState: DirectDebitState = {
  isPopUpOpen: false,
  sortCodeArray: Array(3).fill(''),
  bankAccountNumber: '',
  accountHolderName: '',
  accountHolderConfirmation: false,
  formError: '',
};


// ----- Reducers ----- //

const directDebitReducer = (
  state: DirectDebitState = initialState,
  action: Action,
): DirectDebitState => {

  switch (action.type) {

    case 'DIRECT_DEBIT_POP_UP_OPEN':

      return Object.assign({}, state, {
        isPopUpOpen: true,
      });

    case 'DIRECT_DEBIT_POP_UP_CLOSE':

      return Object.assign({}, state, {
        isPopUpOpen: false,
      });

    case 'DIRECT_DEBIT_UPDATE_SORT_CODE':
      initialState.sortCodeArray[action.index] = action.partialSortCode;
      return Object.assign({}, state, {
        sortCode: initialState.sortCodeArray,
      });

    case 'DIRECT_DEBIT_UPDATE_ACCOUNT_NUMBER':

      return Object.assign({}, state, {
        bankAccountNumber: action.accountNumber,
      });

    case 'DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_NAME':

      return Object.assign({}, state, {
        accountHolderName: action.accountHolderName,
      });

    case 'DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_CONFIRMATION':
      return Object.assign({}, state, {
        accountHolderConfirmation: action.accountHolderConfirmation,
      });

    case 'DIRECT_DEBIT_SET_FORM_ERROR':
      return Object.assign({}, state, {
        formError: action.message,
      });

    case 'DIRECT_DEBIT_RESET_FORM_ERROR': {
      return Object.assign({}, state, {
        formError: '',
      });
    }

    default:
      return state;

  }

};


// ----- Exports ----- //

export { directDebitReducer };
