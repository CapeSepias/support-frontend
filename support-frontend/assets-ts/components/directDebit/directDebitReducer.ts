// ----- Imports ----- //
import { Action, Phase } from './directDebitActions';
// ----- Setup ----- //
export type DirectDebitState = {
	isPopUpOpen: boolean;
	isDDGuaranteeOpen: boolean;
	sortCodeArray: Array<string>;
	sortCodeString: string;
	accountNumber: string;
	accountHolderName: string;
	accountHolderConfirmation: boolean;
	formError: string;
	phase: Phase;
};
const initialState: DirectDebitState = {
	isPopUpOpen: false,
	isDDGuaranteeOpen: false,
	sortCodeArray: Array(3).fill(''),
	sortCodeString: '',
	accountNumber: '',
	accountHolderName: '',
	accountHolderConfirmation: false,
	formError: '',
	phase: 'entry',
};

// ----- Reducers ----- //
const directDebitReducer = (
	state: DirectDebitState = initialState,
	action: Action,
): DirectDebitState => {
	switch (action.type) {
		case 'DIRECT_DEBIT_POP_UP_OPEN':
			return { ...state, isPopUpOpen: true };

		case 'DIRECT_DEBIT_POP_UP_CLOSE':
			return { ...state, isPopUpOpen: false };

		case 'DIRECT_DEBIT_GUARANTEE_OPEN':
			return { ...state, isDDGuaranteeOpen: true };

		case 'DIRECT_DEBIT_GUARANTEE_CLOSE':
			return { ...state, isDDGuaranteeOpen: false };

		case 'DIRECT_DEBIT_UPDATE_SORT_CODE':
			initialState.sortCodeArray[action.index] = action.partialSortCode;
			return { ...state, sortCode: initialState.sortCodeArray };

		case 'DIRECT_DEBIT_UPDATE_SORT_CODE_STRING':
			return { ...state, sortCodeString: action.sortCodeString };

		case 'DIRECT_DEBIT_UPDATE_ACCOUNT_NUMBER':
			return { ...state, accountNumber: action.accountNumber };

		case 'DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_NAME':
			return { ...state, accountHolderName: action.accountHolderName };

		case 'DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_CONFIRMATION':
			return {
				...state,
				accountHolderConfirmation: action.accountHolderConfirmation,
			};

		case 'DIRECT_DEBIT_SET_FORM_ERROR':
			return { ...state, formError: action.message };

		case 'DIRECT_DEBIT_RESET_FORM_ERROR': {
			return { ...state, formError: '' };
		}

		case 'DIRECT_DEBIT_SET_FORM_PHASE': {
			return { ...state, phase: action.phase };
		}

		default:
			return state;
	}
};

// ----- Exports ----- //
export { directDebitReducer };
