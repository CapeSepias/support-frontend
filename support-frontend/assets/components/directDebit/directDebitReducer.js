// @flow

// ----- Imports ----- //

import type { Action, Phase } from './directDebitActions';

// ----- Setup ----- //

export type DirectDebitState = {
  sortCode: string,
  accountNumber: string,
  accountHolderName: string,
  formError: string,
  phase: Phase
};


const initialState: DirectDebitState = {
  sortCode: '',
  accountNumber: '',
  accountHolderName: '',
  formError: '',
  phase: 'entry',
};


// ----- Reducers ----- //

const directDebitReducer = (
  state: DirectDebitState = initialState,
  action: Action,
): DirectDebitState => {

  switch (action.type) {

    case 'DIRECT_DEBIT_UPDATE_SORT_CODE':
      initialState.sortCode = action.sortCode;
      return Object.assign({}, state, {
        sortCode: initialState.sortCode,
      });

    case 'DIRECT_DEBIT_UPDATE_ACCOUNT_NUMBER':

      return Object.assign({}, state, {
        accountNumber: action.accountNumber,
      });

    case 'DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_NAME':

      return Object.assign({}, state, {
        accountHolderName: action.accountHolderName,
      });

    case 'DIRECT_DEBIT_SET_FORM_ERROR':
      return Object.assign({}, state, {
        formError: action.message,
      });

    case 'DIRECT_DEBIT_RESET_FORM_ERROR': {
      return Object.assign({}, state, {
        formError: '',
      });
    }

    case 'DIRECT_DEBIT_SET_FORM_PHASE': {
      return Object.assign({}, state, {
        phase: action.phase,
      });
    }

    default:
      return state;

  }

};


// ----- Exports ----- //

export { directDebitReducer };
