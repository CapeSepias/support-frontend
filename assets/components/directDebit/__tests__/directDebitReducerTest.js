import { directDebitReducer as reducer } from '../directDebitReducer';

// ----- Tests ----- //

describe('direct debit reducer tests', () => {

  it('should return the initial state', () => {
    expect(reducer(undefined, {})).toMatchSnapshot();
  });

  it('should handle DIRECT_DEBIT_POP_UP_OPEN', () => {
    const action = {
      type: 'DIRECT_DEBIT_POP_UP_OPEN',
    };

    const newState = reducer(undefined, action);
    expect(newState.isPopUpOpen).toEqual(true);
  });

  it('should handle DIRECT_DEBIT_POP_UP_CLOSE', () => {
    const action1 = {
      type: 'DIRECT_DEBIT_POP_UP_OPEN',
    };
    const action2 = {
      type: 'DIRECT_DEBIT_POP_UP_CLOSE',
    };

    let newState = reducer(undefined, action1);
    expect(newState.isPopUpOpen).toEqual(true);

    newState = reducer(newState, action2);
    expect(newState.isPopUpOpen).toEqual(false);
  });

  it('should handle DIRECT_DEBIT_UPDATE_SORT_CODE', () => {
    const sortCode = '123456';
    const action = {
      type: 'DIRECT_DEBIT_UPDATE_SORT_CODE',
      sortCode,
    };

    const newState = reducer(undefined, action);

    expect(newState.bankSortCode).toEqual(sortCode);
  });

  it('should handle DIRECT_DEBIT_UPDATE_ACCOUNT_NUMBER', () => {

    const accountNumber = '12345678910';
    const action = {
      type: 'DIRECT_DEBIT_UPDATE_ACCOUNT_NUMBER',
      accountNumber,
    };

    const newState = reducer(undefined, action);

    expect(newState.bankAccountNumber).toEqual(accountNumber);
  });

  it('should handle DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_NAME', () => {

    const accountHolderName = "John Doe";
    const action = {
      type: 'DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_NAME',
      accountHolderName,
    };

    const newState = reducer(undefined, action);

    expect(newState.accountHolderName).toEqual(accountHolderName);
  });

  it('should handle DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_CONFIRMATION', () => {

    const accountHolderConfirmation = true;
    const action = {
      type: 'DIRECT_DEBIT_UPDATE_ACCOUNT_HOLDER_CONFIRMATION',
      accountHolderConfirmation,
    };

    const newState = reducer(undefined, action);

    expect(newState.accountHolderConfirmation).toEqual(accountHolderConfirmation);
  });
});

