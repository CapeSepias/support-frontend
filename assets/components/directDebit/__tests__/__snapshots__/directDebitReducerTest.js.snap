// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`direct debit reducer tests should return the initial state 1`] = `
Object {
  "accountHolderConfirmation": false,
  "accountHolderName": "",
  "bankAccountNumber": "",
  "bankSortCode": "",
  "isPopUpOpen": false,
}
`;
